fuel-vm/src/interpreter/flow.rs

```diff
@@ -149,138 +161,385 @@ where
+    if bytes_read as Word != frame.code_size() {
+        return Err(PanicReason::ContractNotFound.into());
+    }
 
-        let a = self
-            .registers
-            .get(ra.to_u8() as usize)
-            .copied()
-            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, M))?;
+    if frame.code_size_padding() > 0 {
+        let mut padding_range = code_mem_range;
+        padding_range.grow_start(CallFrame::serialized_size() + frame.code_size() as usize);
+        padding_range.write(memory).fill(0);
+    }
+    Ok(code_frame_range.end() as Word)
+}
 
-        let b = self
-            .registers
-            .get(rb.to_u8() as usize)
-            .copied()
-            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, M))?;
+fn call_frame<S>(
+    registers: [Word; VM_REGISTER_COUNT],
+    storage: &S,
+    call: Call,
+    asset_id: AssetId,
+) -> Result<CallFrame, RuntimeError>
+where
+    S: StorageSize<ContractsRawCode> + ?Sized,
+    <S as StorageInspect<ContractsRawCode>>::Error: Into<std::io::Error>,
+{
+    let (to, a, b) = call.into_inner();
 
-        let c = self
-            .registers
-            .get(rc.to_u8() as usize)
-            .copied()
-            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, M))?;
+    let code_size = contract_size(storage, &to)?;
 
-        let d = self
-            .registers
-            .get(rd.to_u8() as usize)
-            .copied()
-            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, M))?;
+    let frame = CallFrame::new(to, asset_id, registers, code_size, a, b);
 
-        self._prepare_call(a, b, c, d)
+    Ok(frame)
+}
+
+impl<'a> From<&'a PrepareCallRegisters<'_>> for ReadRegistersRef<'a> {
+    fn from(registers: &'a PrepareCallRegisters) -> Self {
+        Self {
+            hp: registers.read_registers.hp,
+            sp: registers.read_registers.sp.as_ref(),
+            ssp: registers.read_registers.ssp.as_ref(),
+            fp: registers.read_registers.fp.as_ref(),
+            pc: registers.read_registers.pc.as_ref(),
+            is: registers.read_registers.is.as_ref(),
+            bal: registers.read_registers.bal.as_ref(),
+            cgas: registers.read_registers.cgas.as_ref(),
+            ggas: registers.read_registers.ggas.as_ref(),
+            zero: registers.unused_registers.zero,
+            one: registers.unused_registers.one,
+            of: registers.unused_registers.of,
+            err: registers.unused_registers.err,
+            ret: registers.unused_registers.ret,
+            retl: registers.unused_registers.retl,
+            flag: registers.unused_registers.flag,
+        }
     }
+}
 
-    pub(crate) fn call(&mut self, a: Word, b: Word, c: Word, d: Word) -> Result<ProgramState, RuntimeError> {
-        self._prepare_call(a, b, c, d)?;
-        self.run_call()
+impl<'reg> From<&'reg mut [Word; VM_REGISTER_COUNT]> for PrepareCallRegisters<'reg> {
```

Comment:
[nit] Maybe we want to have some alias for `[Word; VM_REGISTER_COUNT]`

Reply:
I don't want to alias this as its very clear what it is from the type.